posts/inside-rust/2021-03-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,201 @@
+
+If we decide to prefer the latter then we would want to prioritize exposing every error type in the chain of errors to ensure users can always correctly downcast to specific error variants they want to handle. In this situation the guidance would be: _Always return source errors via `Error::source`._
+
+After some discussion we concluded that the reporting focused viewpoint was the ideal one. We prefer splitting our errors into multiple error messages with no duplication of information as opposed to having each error print itself followed by all others. This split moves the responsibility of formatting to the end user, rather than relying on many library authors concatenating error messages together with a consistent format.
+
+This gives us more flexibility and consistency in how we render error reports, and it doesn't prevent users from structuring their error types so all sources are still correctly exposed for downcasting. If we prioritize downcasting however we have no way of untangling the error messages attached to each source.
+
+With all of that said however, we're not in a great position to make this guidance which brings us to the first major problem we're planning on fixing.
+
+## Problem #1: The Missing Reporter
+
+The error `Trait` as described above is only half of a complete error reporting solution. The existence of an interface for reporting implies the existance of a reporter who uses that interface and outputs a fully formatted error report. The problem is, we don't provide any sort of reporter for actually formatting error reports. The expectation is you have to bring your own reporter, either something like `eyre::Report` or a loop printing each error message.
+
+### Solution: Add a new `{:!}` error reporting fmt specifier
+
+Our currently preferred solution to this problem is to build some simple reporting logic into the language's `fmt` grammar. `{:!}` would report types via the `Error` trait rather than via `Display` or `Debug`, and would implicitly iterate over sources and print them all sequentially.
+
+We want to use `fmt::Formatter` flags to customize the output. Calling `println!("Error: {:!}", error);` would output something like this:
+
+```
+Error: error message: source's error message: source's source's ...
+```
+
+Whereas calling `println!("Error: {:#!}", error);` could print something like this (assuming `error` captured a `Backtrace`):
+
+```
+Error:
+   0: error message
+   1: source's error message
+   2: source's source's ...
+
+Backtrace:
+   ...
+```
+
+This should provide reasonable defaults for simple error reporting cases that are encountered most of the time. Applications that need a more complex error report format can then provide their own reporter by wrapping errors in a `Display` type the same way the `Path::display` method does for paths.
+
+## Panics cannot be made from `dyn Error`s, and instead stringifies the error
```

Comment:
The first heading starts with `Problem #1:`, but then no `Problem #2` heading follows in the headings, so I would either remove it from the first heading or add it to the following.